---
title: "![Just Bash It](Figures/BashFigure1.jpeg)"
output:
  html_document
  #github_document:
  #  toc: true
  #  toc_depth: 2
#always_allow_html: TRUE
author: "Center for Health Data Science, November 2023"
---




```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---




## What is this document?

We show here partial solutions to the exercises. We have omitted to show solutions for viewing, moving and creating files as this depends on what your current working directory is, as well as the folder structure you build in the second exercise. 

Also consult the presentations markdown to see the commands we have shown during the course.

&nbsp;

## Exercise 1: Navigating Files and Directories


2. Listing Files

List with latest updated file shown on top:
```{bash, engine="sh", eval=FALSE}
ls -lt
```

List with largest file shown on top:
```{bash, engine="sh", eval=FALSE}
ls -lhS
```

  List hidden files:
```{bash, engine="sh", eval=FALSE}
ls -a
```

The presentations markdown shows some more options to `ls`.
For other tasks we refer to the slideshow. 


---   

&nbsp;


## Exercise 2: File operations mv, cp and rm

4. Remove the whole `TEMP` directory including the file within it.
```{bash, engine="sh", eval=FALSE}
rm -r TEMP
```

For other tasks we refer to the slideshow. 

&nbsp;


---   

&nbsp;

## Exercise 3: Project Organization

1. Make a structured `projects` directory. Really depends on your setup but will involve a lot of `mkdir` and possible `mkdir -p` if you want to create top and sub directories in one go.

3. How large (in bytes and disk space) are the data files you moved from the course materials?
Size in bytes:
```{bash, engine="sh", eval=FALSE}
ls -lh
```

Size in disk space:
```{bash, engine="sh", eval=FALSE}
du -h *
```

4. Check the permissions
```{bash, engine="sh", eval=FALSE}
ls -l
```


5. Make `Readme` file
```{bash, engine="sh", eval=FALSE}
touch Readme.txt
```

---   

&nbsp;


## Exercise 4: Viewing and editing files

3. Expand the `GCF_genomicAnnotation.gff.gz` annotation file while making sure you keep a copy of the compressed version.  
```{bash, engine="sh", eval=FALSE}
gunzip -k GCF_genomicAnnotation.gff.gz
```

5. Get the head of `Annotation.gff` and save it to a new file.
```{bash, engine="sh", eval=FALSE}
head GCF_genomicAnnotation.gff > gcf_head.txt
```

7. Look at the content of one of the two `fq.gz` files with RNA sequencing read without expanding the file
```{bash, engine="sh", eval=FALSE}
zless SRR4420293_R2.fastq.gz
```

Press `q` to leave `zless`.


---   

&nbsp;


## Exercise 5: Data Wrangling 1 


1. Figure out how many, lines, words and characters the `Annotation.gff` file contains using `wc`.
```{bash, engine="sh", eval=FALSE}
wc Annotation.gff
```

2. As we are not interested in the header lines, denoted by hastages, remove these from the `Annotation.gff`and name the new file `Annotation_tmp.gff`. To do this you can employ the command `sed`, see ***pseudo code*** below. You will need to figure out the line numbers of the first line and last line to remove. What do you think the `d` refers to? 

```{bash, engine="sh", eval=FALSE}
sed '1,8d' Annotation.gff > Annotation_tmp.gff
```

3. Use the command `cut` to make a new file which contains only `fields 3,4,5` (i.e. columns with region, start and end) from the `Annotation_tmp.gff` and name it whatever you'd like. To do this you need to set the *tab* delimiter `'\t'`.    
Check that the file looks correct. 
```{bash, engine="sh", eval=FALSE}
cut -d$'\t' -f 3-5 Annotation_tmp.gff > cols345
```

We also would like to have the gene names. They are in field 7  in the `Annotation_tmp.gff`, but field 7 contains itself several entries ids, gene name, locus, tags etc. separated by a `;`. We refer to such separating characters as `delimiters`. One could say that field 7 contains sub-fields delimited by `;`. 

4. We want to get the sub field starting with `gene=`, this is the gene name. To extract this information we will use cut twice. First, we will cut out column 7 from `Annotation_tmp.gff` and name that temporary file something. Then, we cut the temporary file to get field 5
```{bash, engine="sh", eval=FALSE}
cut -d$'\t' -f 7 Annotation_tmp.gff > col7.tmp
cut -d ';' -f 5 col7.tmp > gene_names
```

5. We would like to remove the repetitive `gene=` in each line to just obtain the clean **gene name**. One way to get this is to do a search and replace with `sed `. The syntax is shown as ***pseudo code*** below. The `sed ` command requires you to specify what pattern to match `[pattern to match]` and what pattern to replace it with `[Replace with]`. The patterns are separated by slashes and the whole expression is encased in quotes `'`.

```{bash, engine="sh", eval=FALSE}
sed 's/gene=//g' gene_names > gene_names_clean
```

---   

&nbsp;


## Exercise 6: Data Wrangling 2


1. Pasting files together. 

```{bash, engine="sh", eval=FALSE}
paste -d '\t' cols.tmp gene_names_clean > Annotation_Gene.gff
```

3. Let's have a look at the content of your final file.   

  * Are there any gene sequences with length 0 in the annotation file?
```{bash, engine="sh", eval=FALSE}
sort -t$'\t' -k 5 -n Annotation_Gene_Len.gff > short_first
head short_first #look at the resulting sorted file
```

  * What is the name of the gene with the longest sequence, annotated in your file? - Try the command `sort` with flags  `-k 5 -nr`  
```{bash, engine="sh", eval=FALSE}
sort -t$'\t' -k 5 -nr Annotation_Gene_Len.gff > long_first
head long_first #look at the resulting sorted file
```

  * Does our organism of study, Arabidopsis Thaliana, have the `TERT` gene? - Try the command `grep`.
```{bash, engine="sh", eval=FALSE}  
grep 'TERT' Annotation_Gene_Len.gff
```
  
  * All living organisms have polymerase genes, including Arabidopsis Thaliana. How many types of `POL` genes are annotated?
```{bash, engine="sh", eval=FALSE}  
grep 'POL' Annotation_Gene_Len.gff
```

  * How many gene annotation lines in the file pertain to transfer RNA (`tRNA`)? - Try the command `grep` with flag `-c`.
```{bash, engine="sh", eval=FALSE}  
grep 'tRNA' Annotation_Gene_Len.gff -c
```

---   

&nbsp;

4. Adding the header
```{bash, engine="sh", eval=FALSE}  
#create the header with nano or other editor
cat header.gff Annotation_Gene_Len.gff > Annotation_21Nov2022.gff
mv Annotation_21Nov2022.gff ../Results
```


---   

&nbsp;


## Exercise 7: Redirection & Pipes

2. Find the types of genomic features. 

```{bash, engine="sh", eval=FALSE}  
sed '1,8d' Annotation.gff | cut -f 3 -d$'\t' | sort | uniq
```

3. Re-run the command line you used above, but this time redirect the output directly to the `Data/Generated` directory by specifying the path and the name you would like the output file to have. Check the `Data/Generated` directory to check that you have correctly made the file.
```{bash, engine="sh", eval=FALSE}  
sed '1,8d' Annotation.gff | cut -f 3 -d$'\t' | sort | uniq > ../Data/Generated/feature_types.txt
```

4. Using a single command line, figure out the gene name of the microRNA (miRNA) gene which has the smallest genomic starting coordinate.

```{bash, engine="sh", eval=FALSE}  
grep 'MIR' Annotation.gff | sort -k 4 -n -t$'\t'| cut -f 7 -d$'\t' | cut -f 5 -d ';' | head
```
 
The microRNA (miRNA) gene which has the smallest genomic starting coordinate is `gene=MIR5642b` starting at `2805`. You can see the coordinate by omitting the two cut statements so you see the full line.  

5. View the last 20 commands from your history. You need tail to look at the end of the file. The first command would be at the top.

```{bash, engine="sh", eval=FALSE}  
history | tail -n 20
```

6. Redo **points 2-6 from Exercise 4 (parts 1 & 2) above**. You should end up with a file containing the 4 columns: `Region, Start, End and GeneName (only)`. To archive this, chain together commands from points 2-6. 

Assuming your starting file is `GCF_genomicAnnotation.gff.gz` unzipped into `Annotation.gff` this works:
```{bash, engine="sh", eval=FALSE}

sed '1,8d' Annotation.gff | cut -d$'\t' -f 7 | cut -d ';' -f 5 | sed 's/gene=/''/g' > geneName.gff 
sed '1,8d' Annotation.gff| cut -f 3-5 | paste -d '\t' - geneName.gff >  output.gff

```

The solution proposed here is not the only valid solution, merely an example. If you have a different solution that works that's fine too.

---   

&nbsp;


## Exercise 8: Part 1: Shell Scripts 

A working script could look like this:

```{bash, engine="sh", eval=FALSE}

#assuming your project folder is 'my_project'  and the unzipped gff should be in the scratch folder
gunzip -c $1 > ~/my_project/Scratch/unzipped.gff

#change into the scratch folder. We will produce all our temporary files here
cd ~/my_project/Scratch/

sed '1,8d' unzipped.gff | cut -d$'\t' -f 7 | cut -d ';' -f 5 | sed 's/gene=/''/g' > geneName.gff 
sed '1,8d' unzipped.gff | cut -f 3-5 | paste -d '\t' - geneName.gff >  ~/my_project/Data/output.gff

rm geneName.gff
rm unzipped.gff

echo 'Done!'

```


We have also uploaded it as a file, ex8.sh. 

To run it
```{bash, engine="sh", eval=FALSE}
bash -x ex8.sh ~/my_project/Data/ecoli.gff.gz
```

---   

&nbsp;


## Exercise 8: Part 2: Loops

1. Make a short script with a for loop that does the following: 

* Go to the `Data` folder
* For each zipped file, i.e. each file ending with .gz:
* Display the file name
* Display the first 10 lines without unzipping the entire file.

```{bash, engine="sh", eval=FALSE}
cd Data
for file in *.gz
do
  echo $file
  gunzip -k $file | head
done

```



---   

&nbsp;

## Exercise 9: Software Installation, Upkeep & More 

Launch Firefox:
```{bash, engine="sh", eval=FALSE}
open -a Firefox

or 

start Firefox
```

Find process and `kill` it:
```{bash, engine="sh", eval=FALSE}
ps aux | grep Firefox

# Kill application
kill -9 [pid]
```


Install `fastqc`:

```{bash, engine="sh", eval=FALSE}
brew install fastqc

or 

apt-get fastqc
```

Move fastq files:
```{bash, engine="sh", eval=FALSE}
mv *.fastq.gz 
```

Run `fastqc`:
```{bash, engine="sh", eval=FALSE}
fastqc SRR4420293_R1.fastq.gz SRR4420293_R2.fastq.gz -o [~/mypath]/Results
```

If you to provide the fullpath you can do it like so:
```{bash, engine="sh", eval=FALSE}
~/Downloads/fastqc_v0.12.1/FastQC/fastqc SRR4420293_R1.fastq.gz SRR4420293_R2.fastq.gz -o [~/mypath]/Results
```

