---
title: "![](BashFigure1.jpeg)"
output:
  github_document:
    toc: true
    toc_depth: 2
always_allow_html: TRUE
author: "Center for Health Data Science, July 2022"
---




```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---


## Introduction

This markdown document contains of exercises for the introductory workshop on command line use, entitled `Just Bash It`, developed and hosted by Center for Health Data Science (HeaDS) at the Faculty of Health and Medical Sciences ([HeaDS](https://heads.ku.dk/)), in collaboration with the University Library ([KUB](https://kub.kb.dk/datalab)) Data Lab.   
This one day workshop is targeted at biomedical and health researchers at the University of Copenhagen with no prior experience in bash command line use.   

The workshop consists of a slideshow with slides to each section, in addition to hands-on presentations (code-along) and accompanying exercises. 


## Exercise 1: Navigating Files and Directories


## Exercise 2: Project Organization

Let's get structured!   

1. Make a `projects` directory at `/home/user/Desktop` on your computer with all the sub-folders shown on **slide 15** in the slideshow. You are free to name the project within the  `projects` folder whatever you would like, e.g. `Just_Bash_It, Intro_to_command_line, First_Project, ...`.   
**Cheat Sheet 1** in the slideshow will have the commands you need. 

You should have downloaded the course materials including the raw data files to you computer, maybe it is in your downloads folder or perhaps you moved it somewhere else. 

2. Using the command-line, navigate to where you downloaded the course materials an go to the directory named `Data`, here you should see three files, all with the extension `.gz`. Move these three files to the project directory you have made and place them in the correct sub-folder.

3. How large are the data files you moved from the course materials?

4. Check the permissions of the data files from the course materials. Do you have permission to `read`, `write` and `execute` them? If you are not allowed to execute the, what is the reason for this?

5. Make a new file called `readme` within your project directory.


## Exercise 3: Working with Files and Directories

1. Go to your project directory and edit the `readme` file you made in _Exercise 2_ using the `nano` editor. A `readme` file should contain information about what a certain directory contains, its purpose and who owns it/is the editor.

2. Move to the sub-directory `Data/Raw` here you should have the three data files you moved from the course material directory.

Two of these files are `fastq` files, they contain RNAseq reads from a bulk-RNA sequencing experiment using Arabidopsis thaliana (thale cress). The extension `_1` and `_2` denote that reads are paired-end (e.g. read 1 and read 2). The third file, `gff`, contains annotations of genes and other genomic feature from the organism of study, Arabidopsis thaliana.

3. Expand the `.gff.gz` annotation file and make sure you keep the compressed version.   
**HINT:** look into what flags/arguments need to be specified when decompressing). 

4. Try out the commands `less, cat, head, tail` on the expanded `.gff` file. Note what it contains. 

5. Look at the content of one of the two `fq.gz` files with RNA sequencing read, **N.B.** this time without expanding the file!


## Exercise 4: Commands, Redirection & Pipes


## Exercise 5: Advanced Commands


## Exercise 6: Editors and Shell scripts 


## Exercise 7: Loops and Workflow Language 


## Exercise 8: Software Installation, Upkeep & More

